### ISSUE TITLE
[router/testing] `No base href set` errors in tests appear with router 3.1.0

### ISSUE THREAD

**Issue by cexbrayat** [angular/angular#12295]
**I'm submitting a ...**  (check one with "x")

```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

Given a module using the RouterModule:

```typescript
@NgModule({
  declarations: [
    AppComponent,
    UsersComponent,
    UserComponent,
    HomeComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    RouterModule.forRoot([
      { path: '', component: HomeComponent},
      { path: 'users', component: UsersComponent}
    ])
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

`UsersComponent` uses `UserComponent` and a `RouterLink` in its template:

```html
<app-user></app-user>
<a routerLink="/">Home</a>
```

 whereas `UserComponent` is just plain HTML for this repro.

Testing a simple component using Angular 2.1.0 and Router ~3.0.0 
for a component using `routerLink` in its template like `UsersComponent`, we had to add `RouterTestingModule` in the imports of the testing module: 

```typescript
import { TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { AppModule } from '../app.module';
import { UsersComponent } from './users.component';

describe('Component: Users', () => {
  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [AppModule, RouterTestingModule],
    });
  });

  it('should create the component', () => {
    let fixture = TestBed.createComponent(UsersComponent);
    let app = fixture.debugElement.componentInstance;
    expect(app).toBeTruthy();
  });
});
```

whereas for a "dumb" component like `UserComponent` we did not:

```typescript
import { TestBed } from '@angular/core/testing';
import { AppModule } from '../app.module';
import { UserComponent } from './user.component';

describe('Component: User', () => {
  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [AppModule],
    });
  });

  it('should create the component', () => {
    let fixture = TestBed.createComponent(UserComponent);
    let app = fixture.debugElement.componentInstance;
    expect(app).toBeTruthy();
  });
});
```

**Now, using the router 3.1.0, the above test throws: 
`No base href set. Please provide a value for the APP_BASE_HREF token or add a base element to the document.`**

This can be fixed by adding the `RouterTestingModule` in the imports of the testing module for the dumb component too, but this is a regression/breaking change compared to the behavior we used to have.

**Expected behavior**

We should be able to test a component that does not rely on `routerLink` without having to import the `RouterTestingModule` as we used to.

**Minimal reproduction of the problem with instructions**
See above

**Please tell us about your environment:**

MacOSX
- **Angular version:** 2.0.X
  Angular 2.1.0, Router 3.1.0
- **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
  All
- **Language:** [all | TypeScript X.X | ES6/7 | ES5]
  All
- **Node (for AoT issues):** `node --version` =

**Comment by DzmitryShylovich** [comment]
@cexbrayat still an issue? could you add a small demo on Plunkr?

**Comment by cexbrayat** [comment]
Thank you for looking into this.

This is still an issue with 3.4.3. I don't know how to setup tests on plunkr, but here is a repo with repro:
https://github.com/cexbrayat/12295-repro (angular-cli app).

The last commit removes `RouterTestingModule` in a test where it should not be needed. Run `ng test` and you'll have the `No base href set. Please provide a value for the APP_BASE_HREF token or add a base element to the document.`  error.
Add `RouterTestingModule` again, and the tests will be fine.

**Comment by DaveX81** [comment]
I havethe same problem but i couldnt fix that.

I created an angular2 app with angular cli. 
If i generate a component and use it on my app component:
`<app-mycomponent></app-mycomponent>`
It is working if import the RouterTestingModule and declare my component in the app spec.

But if i create a module and export this component and import this module in my spec there is no workaround.
`@NgModule({
  imports: [
    CommonModule,
    AppRoutingModule
  ],
  declarations: [
    HeaderComponent,
    MaincontentComponent,
    FooterComponent,
    LayoutComponent
  ],
  exports: [
    LayoutComponent
  ]
})
export class LayoutModule { }`

`describe('AppComponent', () => {
  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        LayoutModule
      ],
      declarations: [
        AppComponent
      ]
    });
    TestBed.compileComponents();
  });`

This will result in:

> hrome 55.0.2883 (Windows 7 0.0.0) AppComponent should create the app FAILED
	Failed: No base href set. Please provide a value for the APP_BASE_HREF token or add a base element to the document.
	Error: No base href set. Please provide a value for the APP_BASE_HREF token or add a base element to the document.

if i change my app.component.spec.ts by declaring all components directly it is working:
`describe('AppComponent', () => {
  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        AuthenticationModule,
        RouterTestingModule
      ],
      declarations: [
        AppComponent
        LayoutComponent,
        HeaderComponent,
        FooterComponent,
        MaincontentComponent
      ]
    });
    TestBed.compileComponents();
  });`

**Comment by bhargav-sae** [comment]
- When I tried to include module in `TestBed.configureTestingModule`,
- With same configuration, when I run other test without including any module, it runs fine.

```
Error: No base href set. Please provide a value for the APP_BASE_HREF token or add a base element to the document.
        at new PathLocationStrategy (webpack:///~/@angular/common/bundles/common.umd.js:525:0 <- config/karma-shim.js:25939:24)
        at provideLocationStrategy (webpack:///~/@angular/router/bundles/router.umd.js:4019:0 <- config/karma-shim.js:58320:14)
        at NgModuleInjector.get (/DynamicTestModule/module.ngfactory.js:190:77)
        at NgModuleInjector.get (/DynamicTestModule/module.ngfactory.js:195:84)
        at NgModuleInjector.get (/DynamicTestModule/module.ngfactory.js:279:151)
        at NgModuleInjector.DynamicTestModuleInjector.createInternal (/DynamicTestModule/module.ngfactory.js:422:61)
        at NgModuleInjector.create (webpack:///~/@angular/core/bundles/core.umd.js:6715:57 <- config/karma-shim.js:18704:81)
        at NgModuleFactory.create (webpack:///~/@angular/core/bundles/core.umd.js:6700:0 <- config/karma-shim.js:18689:23)
        at TestBed._initIfNeeded (webpack:///~/@angular/core/bundles/core-testing.umd.js:771:0 <- config/karma-shim.js:11679:69)
        at TestBed.createComponent (webpack:///~/@angular/core/bundles/core-testing.umd.js:840:0 <- config/karma-shim.js:11748:19)
        at Function.TestBed.createComponent (webpack:///~/@angular/core/bundles/core-testing.umd.js:670:0 <- config/karma-shim.js:11578:34)
        at Object.<anonymous> (webpack:///src/app/home/home.component.spec.ts:42:23 <- config/karma-shim.js:28408:38)
        at ZoneDelegate.invoke (webpack:///~/zone.js/dist/zone.js:232:0 <- config/karma-shim.js:8612:27)
        at ProxyZoneSpec.onInvoke (webpack:///~/zone.js/dist/proxy.js:79:0 <- config/karma-shim.js:10241:40)
        at ZoneDelegate.invoke (webpack:///~/zone.js/dist/zone.js:231:0 <- config/karma-shim.js:8611:33)
        at Zone.run (webpack:///~/zone.js/dist/zone.js:114:0 <- config/karma-shim.js:8494:44)
        at Object.<anonymous> (webpack:///~/zone.js/dist/jasmine-patch.js:102:0 <- config/karma-shim.js:10447:35)
        at ZoneQueueRunner.jasmine.QueueRunner.ZoneQueueRunner.execute (webpack:///~/zone.js/dist/jasmine-patch.js:132:0 <- config/karma-shim.js:10477:43)
        at ZoneQueueRunner.jasmine.QueueRunner.ZoneQueueRunner.execute (webpack:///~/zone.js/dist/jasmine-patch.js:132:0 <- config/karma-shim.js:10477:43)
        at ZoneDelegate.invokeTask (webpack:///~/zone.js/dist/zone.js:265:0 <- config/karma-shim.js:8645:36)
        at Zone.runTask (webpack:///~/zone.js/dist/zone.js:154:0 <- config/karma-shim.js:8534:48)
        at drainMicroTaskQueue (webpack:///~/zone.js/dist/zone.js:401:0 <- config/karma-shim.js:8781:36)
```

**Comment by Draccoz** [comment]
Any update on that issue? It prevents testing without code duplication (module declarations) and is a huge pain. Please assign a higher priority to that issue...

**Comment by tylerlindell** [comment]
@Draccoz, you nailed that.  thank you `providers: [{provide: APP_BASE_HREF, useValue: '/'}]`

for future reference: `APP_BASE_HREF` comes from `import { APP_BASE_HREF } from '@angular/common';`

**Comment by richardsengers** [comment]
So....how could someone knows this?
Where is this documented?Or is there an easy way to understand how this exactly works?

**Comment by Atis0505** [comment]
If you gonna use this, it's gonna solve the problem in angular 4.
<head>
    <base href="/">
</head>

**Comment by zuosc** [comment]
ASPNETCORE ANGULAR5.1

add  '     <base href="~/" />'  in _Layout.cshtml

like this:

```
    <base href="~/" />
    <link rel="stylesheet" href="~/dist/vendor.css" asp-append-version="true" />
```

**Comment by gpresland** [comment]
I was receiving this in Angular 6 after following the tutorial from:

https://angular.io/tutorial/toh-pt5

The solution was to add the `APP_BASE_HREF` described here:

https://angular.io/api/common/APP_BASE_HREF

in your _app-routing.module.ts_ file.

**Comment by winfredchen** [comment]
I got it when try to update to Angular 6

Fixed by put the <app-root> tag in to body tag in index.html (the default index.html doesn't have a body tag)

```
<body>
    <app-root>
        ...
    </app-root>
</body>
```

**Comment by cexbrayat** [comment]
I'm going to close the issue, as my PR on angular/cli has been merged, and released in `7.1.0-beta.1`. The error should not appear in unit tests anymore.

See https://github.com/angular/angular-cli/pull/12889 for more details
